### PR TITLE
MAINT: more scalar array conversion fixes for optimize

### PR DESCRIPTION
**Pull request by nschloe**
Follow-up on https://github.com/scipy/scipy/pull/13864.

This PR fixes the rest of the scalar-array conversions in scipy.optimize. (There are a handful of other instances in the rest of scipy, I'll put those in a separate PR.)

### CODE REVIEW

File: scipy/optimize/_differentialevolution.py [hunk 1]
```diff
@@ -1343,7 +1343,17 @@ def __init__(self, f, args):
         self.args = [] if args is None else args
 
     def __call__(self, x):
-        return self.f(x, *self.args)
+        fx = self.f(np.copy(x), *self.args)
+        # Make sure the function returns a true scalar
+        if not np.isscalar(fx):
```
Comment (andyfaff) [inline scipy/optimize/_differentialevolution.py:1348]: With the way that the minimisation is done it should be ultimately possible to work with a vectorised `func`, which would then return a vector as well. Thus a future API change could be removing this code anyway.

It shouldn't matter if the user function returns `1, [1], np.array([[[[1]]]])` as the return from `func` is made into an array and squeezed on L901.

I think before the changes in L1347 to L1356 are committed a test should be made to see what happens when a length > 1 array is returned from `func`. I'm guessing that something will go wrong in L901, possibly falling into the except clause on L902.

It doesn't make sense to check for 'scalar' nature here, it just adds too much overhead. Can you make the test on L901 in `_calculate_population_energies`:
```
calc_energies = np.squeeze(calc_energies)
if calc_energies.size != nfevs:
    raise ValueError(
        "The user-provided objective function must return a scalar value."
    )
energies[0:nfevs] = calc_energies
```
Note that `_calculate_population_energies` is *always* called at the start of a solve cycle (L1057).

**Review by andyfaff — CHANGES_REQUESTED** [review]
Please make the requested changes to differential_evolution. Renaming `_FunctionWrapper` is fine, but the other changes need to be modified a bit.

File: scipy/optimize/_differentialevolution.py (hunk 1 above)
Comment (nschloe) [reply scipy/optimize/_differentialevolution.py:1348]: > With the way that the minimisation is done it should be ultimately possible to work with a vectorised func, which would then return a vector as well. 

Cool! How would that work with `minimize(self.func)` (L903)? (That was the line that had me assume you always have to have a scalar-valued function.)

Comment (nschloe) [reply scipy/optimize/_differentialevolution.py]: Never mind. I just added `np.squeeze` in another line which also fixes the problem.

**Comment by nschloe** [comment]
@andyfaff I made the changes. (Not sure now if I had already pinged you.) There's one test failing, but since it's just one I would assume the failure is not related to the changes in this PR. How can I restart it?

**Comment by nschloe** [comment]
Let me know if there's anything I can do to improve this PR.

**Comment by andyfaff** [comment]
This still isn't right. The change you made to differential evolution isn't of the type that is needed. With the following example only `f2` gives issues:

```
import numpy as np
from scipy.optimize import differential_evolution

f0 = lambda x: np.sum(x**2)                           # returns a scalar
f1 = lambda x: np.array([np.sum(x**2)])               # returns length 1 array
f2 = lambda x: np.array([np.sum(x**2), np.sum(x)])    # returns a length 2 array

differential_evolution(f0, [(-2, 2), (-2, 2)])
differential_evolution(f1, [(-2, 2), (-2, 2)])
differential_evolution(f2, [(-2, 2), (-2, 2)])
```

No changes need to be made for returns f0 or f1. The problem with f2 is already raised as:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/miniconda3/envs/dev3/lib/python3.8/site-packages/scipy/optimize/_differentialevolution.py in _calculate_population_energies(self, population)
    967                                                   parameters_pop[0:nfevs]))
--> 968             energies[0:nfevs] = np.squeeze(calc_energies)
    969         except (TypeError, ValueError) as e:

ValueError: could not broadcast input array from shape (30,2) into shape (30,)

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
/var/folders/m8/gwtcncws12jf60xw5n6knwnw0000gn/T/ipykernel_53058/2090156689.py in <module>
      1 differential_evolution(f0, [(-2, 2), (-2, 2)])
      2 differential_evolution(f1, [(-2, 2), (-2, 2)])
----> 3 differential_evolution(f2, [(-2, 2), (-2, 2)])

~/miniconda3/envs/dev3/lib/python3.8/site-packages/scipy/optimize/_differentialevolution.py in differential_evolution(func, bounds, args, strategy, maxiter, popsize, tol, mutation, recombination, seed, callback, disp, polish, init, atol, updating, workers, constraints, x0)
    327                                      constraints=constraints,
    328                                      x0=x0) as solver:
--> 329         ret = solver.solve()
    330 
    331     return ret

~/miniconda3/envs/dev3/lib/python3.8/site-packages/scipy/optimize/_differentialevolution.py in solve(self)
    841             # only work out population energies for feasible solutions
    842             self.population_energies[self.feasible] = (
--> 843                 self._calculate_population_energies(
    844                     self.population[self.feasible]))
    845 

~/miniconda3/envs/dev3/lib/python3.8/site-packages/scipy/optimize/_differentialevolution.py in _calculate_population_energies(self, population)
    970             # wrong number of arguments for _mapwrapper
    971             # or wrong length returned from the mapper
--> 972             raise RuntimeError(
    973                 "The map-like callable must be of the form f(func, iterable), "
    974                 "returning a sequence of numbers the same length as 'iterable'"

RuntimeError: The map-like callable must be of the form f(func, iterable), returning a sequence of numbers the same length as 'iterable'
```

The main issue with this is that the error message isn't clear on what the user has done wrong.The code in `_calculate_population_energies` can possibly be changed to:

```
try:
    calc_energies = list(self._mapwrapper(self.func,
                                          parameters_pop[0:nfevs]))
    calc_energies = np.squeeze(calc_energies)
except (TypeError, ValueError) as e:
    # wrong number of arguments for _mapwrapper
    # or wrong length returned from the mapper
    raise RuntimeError(
        "The map-like callable must be of the form f(func, iterable), "
        "returning a sequence of numbers the same length as 'iterable'"
    ) from e

if calc_energies.size != nfevs:
    raise RuntimeError("func(x, *args) must return a scalar value")

energies[0:nfevs] = calc_energies
```

If you make this change then please add a test that the `f2` function above raises a RuntimeError with the right message.

**Comment by nschloe** [comment]
Thanks @andyfaff for the response. I've applied your suggested fix to get a better error message. Tests should work too.

File: scipy/optimize/_differentialevolution.py [hunk 2]
```diff
@@ -958,14 +958,15 @@ def _calculate_population_energies(self, population):
         num_members = np.size(population, 0)
         nfevs = min(num_members,
                     self.maxfun - num_members)
```
Comment (andyfaff) [inline scipy/optimize/_differentialevolution.py:960]: I think there is an existing bug here. I think the line should be:
```
nfevs = min(num_members, self.maxfun - self._nfev)
```
Which means the line you added isn't needed.

File: scipy/optimize/_differentialevolution.py [hunk 3]
```diff
@@ -974,6 +975,12 @@ def _calculate_population_energies(self, population):
                 "returning a sequence of numbers the same length as 'iterable'"
             ) from e
 
+        # nfecs can be negative
```
[the remark this answers is not in the document]
Comment (andyfaff) [reply scipy/optimize/_differentialevolution.py:978]: I think that comment can be removed, nfevs should never be negative once the bug is fixed.

**Review by andyfaff — CHANGES_REQUESTED** [review]

**Comment by andyfaff** [comment]
@nschloe `differential_evolution` can be vectorised as follows:

```
import numpy as np
from scipy.optimize import differential_evolution, rosen

def vectorised_rosen(func, x):
    # func needs to return an array that's the same length as the population, so transpose may be needed
    return func(x.T)

# this would be the comparison you needed to do
%timeit differential_evolution(rosen, bounds=[(0, 2), (0, 2)], workers=vectorised_rosen, updating="deferred")
%timeit differential_evolution(rosen, bounds=[(0, 2), (0, 2)], workers=1, updating="deferred")
```

If vectorising the users function can provide good speed benefits, then perhaps we should add the example to the docstring.

File: scipy/optimize/_differentialevolution.py (hunk 2 above)
Comment (nschloe) [reply scipy/optimize/_differentialevolution.py:960]: Should I change it this way?

**Comment by nschloe** [comment]
The code you suggest gives
```
185 ms ± 7.73 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
231 ms ± 9.91 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
so the vectorized version is about 20% faster. Should this be added to the docstring in this PR?

File: scipy/optimize/_differentialevolution.py (hunk 2 above)
Comment (andyfaff) [reply scipy/optimize/_differentialevolution.py:960]: No thank you. Just remove the line you added, I'll make a separate PR.

**Comment by andyfaff** [comment]
We'll add the docstring later

File: scipy/optimize/_differentialevolution.py (hunk 2 above)
Comment (nschloe) [reply scipy/optimize/_differentialevolution.py:960]: The `max` is needed for the `if calc_energies.size != nfevs:` to work, so I cannot remove it.

Comment (andyfaff) [reply scipy/optimize/_differentialevolution.py:960]: I fixed this bug in #14520, so the line `nfevs = max(nfevs, 0)` is not needed.

**Comment by andyfaff** [comment]
Would it be possible for you to make the change I requested, and rebase on master?

**Comment by andyfaff** [comment]
I'm happy for this PR to be merged now. @nschloe could you squash the commits down before merge, 11 is a bit too many for the changes here.
Alternatively I can do a squash and merge?

**Comment by nschloe** [comment]
> Alternatively I can do a squash and merge?

This.

**Comment by nschloe** [comment]
Thanks for the concise review!

**Pull request merged**